ArduCopter/crash_check.cpp

```diff
@@ -53,6 +58,62 @@ void Copter::crash_check()
     }
 }
 
+// check for loss of thrust and trigger thrust boost in motors library
+void Copter::thrust_loss_check()
+{
+    static uint16_t thrust_loss_counter;  // number of iterations vehicle may have been crashed
+
+    // exit immediately if motor failure handling already engaged
+    if (motors->get_thrust_boost()) {
+        return;
+    }
+
+    // return immediately if disarmed
+    if (!motors->armed() || ap.land_complete) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for desired angle over 15 degrees
+    // todo: add thrust angle to AC_AttitudeControl
+    const Vector3f angle_target = attitude_control->get_att_target_euler_cd();
+    if (safe_sqrt(sq(angle_target.x) + sq(angle_target.y)) > THRUST_LOSS_CHECK_ANGLE_DEVIATION_CD) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for throttle over 90% or throttle saturation
+    if ((attitude_control->get_throttle_in() < THRUST_LOSS_CHECK_MINIMUM_THROTTLE) && (!motors->limit.throttle_upper)) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check throttle is over 25%
+    if ((attitude_control->get_throttle_in() < 0.25f)) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // check for decent
+    if (inertial_nav.get_velocity_z() >= 0.0f) {
+        thrust_loss_counter = 0;
+        return;
+    }
+
+    // we may have lost thrust
+    thrust_loss_counter++;
+
+    // check if thrust loss for 1 second
+    if (thrust_loss_counter >= (THRUST_LOSS_CHECK_TRIGGER_SEC * scheduler.get_loop_rate_hz())) {
```

Comment:
this looks like it will flood the log and mavlink

Reply:
// exit immediately if thrust boost is already engaged
    if (motors->get_thrust_boost()) {
        return;
    }

That gets prevented by this.